### PR TITLE
[rocprofiler-compute] Add functional tests for iteration multiplexing with roofline analysis

### PR DESCRIPTION
**Pull request by jamessiddeley-amd**
## Motivation

Iteration multiplexing counter accuracy is tested for general counters (blocks 6, 7, 10, 16) but not for roofline counters (block 4). Also no existing test verifies that the analysis mode roofline pipeline (calc_ai_analyze via 0400_roofline.yaml) works end to end on iteration multiplexed profiling data. This PR does both.

## Technical Details

Integrates roofline (block 4) into already existing iteration multiplexing tests and adds 2 new corner case tests under the `iteration_multiplexing_2` marker.

Modified existing tests:
`test_iteration_multiplexing_deterministic_counter_accuracy` -- added block 4 to the block list, added roofline assertions (roofline.csv validity + rocprof-compute analyze exit code 0) on the `kernel_launch_params` workload. Roofline benchmark only runs on the final profiling pass to avoid unnecessary overhead.
`test_iteration_multiplexing_stochastic_counter_accuracy` -- same pattern which is block 4 added, roofline assertions on the `kernel_launch_params` workload.

2 new tests:
`test_iteration_multiplexing_insufficient_dispatches` -- profiles `laplace_eqn -i 3` (which is below the ~20 replay minimum) with blocks 4 + 6.1.5, verifies `kernels_with_missing_counters` warning in `profiling_config.yaml`, and asserts analyze exits gracefully.
`test_iteration_multiplexing_data_types` -- makes sure `rocprof-compute analyze` works with both `--roofline-data-type` `FP32` and `FP16` on multiplexed data.

## JIRA ID

AIPROFCOMP-172

## Test Plan

Run `pytest tests/test_profile_general.py -m iteration_multiplexing_2` on a system (I've done MI300A)
Run `pytest tests/test_profile_general.py -m iteration_multiplexing_stochastic` 

Verify tests pass
Run ctest and verify it ran successfully

## Test Result

On MI300A:
`python -m pytest -s -k "test_iteration_multiplexing_deterministic_counter_accuracy or test_iteration_multiplexing_stochastic_counter_accuracy or test_iteration_multiplexing_insufficient_dispatches or test_iteration_multiplexing_data_types" tests/test_profile_general.py`

image

ctest also full passes
## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 1]
```diff
@@ -51,6 +51,17 @@
 config["app_hip_dynamic_shared"] = ["./tests/hip_dynamic_shared"]
 config["app_laplace_eqn"] = ["./tests/laplace_eqn", "-i", "5000"]
 config["app_laplace_eqn_iter"] = ["./tests/laplace_eqn", "-i", "15000"]
+config["app_laplace_eqn_insufficient"] = ["./tests/laplace_eqn", "-i", "3", "-o", "0"]
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Dont think you need to provide `-o 0`, is there a reason why we are passing that?

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 2]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
     assert "--set" in output
 
     test_utils.clean_output_dir(config["cleanup"], workload_dir)
+
+
+@pytest.mark.roofline_iteration_multiplexing
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Lets group similar tests together, any new iter mplx tests should go after `test_iteration_multiplexing_stochastic_counter_accuracy()` and not arbitrarily at the end of the file

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 3]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+def test_roofline_iteration_multiplexing_laplace_eqn(
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: We dont need this test, just add `--block 4` to `test_iteration_multiplexing_deterministic_counter_accuracy()` and `test_iteration_multiplexing_stochastic_counter_accuracy()` and move roofline.csv assertions over there.

roofline.csv assertions should come right after pmc_perf.csv assertion, this assertions should be skipped for MI100

You will have to update the aforementioned test cases to accept `binary_handler_analyze_rocprof_compute` fixture and execute analysis and assert pass. This way we will also be checking whether analysis passes for non roofline blocks, which is useful

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 4]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+        workload_dir_mux,
+    ])
+    assert code == 0
+
+    test_utils.clean_output_dir(config["cleanup"], workload_dir_baseline)
+    test_utils.clean_output_dir(config["cleanup"], workload_dir_mux)
+
+
+@pytest.mark.roofline_iteration_multiplexing
+def test_roofline_iteration_multiplexing_vcopy_multikernel(
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: laplace_eqn is already multikernel we dont need this test

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 5]
```diff
@@ -51,6 +51,17 @@
 config["app_hip_dynamic_shared"] = ["./tests/hip_dynamic_shared"]
 config["app_laplace_eqn"] = ["./tests/laplace_eqn", "-i", "5000"]
 config["app_laplace_eqn_iter"] = ["./tests/laplace_eqn", "-i", "15000"]
+config["app_laplace_eqn_insufficient"] = ["./tests/laplace_eqn", "-i", "3", "-o", "0"]
+config["app_vcopy_multikernel_iter"] = [
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py:54]: laplace_eqn is already multi kernel, we dont need this

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 6]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+        "--path",
+        workload_dir_mux,
+    ])
+    assert code == 0
+
+    test_utils.clean_output_dir(config["cleanup"], workload_dir_baseline)
+    test_utils.clean_output_dir(config["cleanup"], workload_dir_mux)
+
+
+@pytest.mark.roofline_iteration_multiplexing
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: This behavior is true for any block and not only roofline. so this test is not roofline specific even though it uses `--block 4`, i would recommend using marker `iteration_multiplexing_2` since this is a general iter. mplx. test not any roofline specific behavior test.

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 7]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+):
+    """
+    Corner case: too few dispatches for iteration multiplexing to cover all
+    counter subsets. Uses laplace_eqn -i 3 -o 0 (3 dispatches, single launch
+    config) which is well below the ~20 replay minimum.
+
+    Verifies profiling warns about missing counters and analysis handles
+    the situation gracefully without crashing.
+    """
+    if soc in ("MI100"):
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: This behavior should be tested for multiple block lets say `--block 4 6.1.5` this should not be a roofline specific tests and hence skip for MI 100 is not necessary

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 8]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+
+
+@pytest.mark.roofline_iteration_multiplexing
+def test_roofline_iteration_multiplexing_insufficient_dispatches(
+    binary_handler_profile_rocprof_compute,
+    binary_handler_analyze_rocprof_compute,
+):
+    """
+    Corner case: too few dispatches for iteration multiplexing to cover all
+    counter subsets. Uses laplace_eqn -i 3 -o 0 (3 dispatches, single launch
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: You dont need `-o 0`, `-i 3` will ensure 3 dispatches

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 9]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+        "kernels_with_missing_counters",
+        f"{workload_dir}/profiling_config.yaml",
+    )
+
+    code = binary_handler_analyze_rocprof_compute([
+        "analyze",
+        "--path",
+        workload_dir,
+    ])
+    assert code in (0, 1)
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Lets assert either exit code 0 or 1, asseting either of them doesnt make sense. Either analyze passes or fail, asserting both doesnt make sense.

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 10]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+    Verifies profiling warns about missing counters and analysis handles
+    the situation gracefully without crashing.
+    """
+    if soc in ("MI100"):
+        pytest.skip("Roofline not supported on MI100")
+        return
+
+    options = [
+        "--block",
+        "4",
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: This should not be a roofline block only tests, please add another block such as 6.1.5 as used in other iter. mplx. tests

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 11]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+        config,
+        workload_dir,
+        options,
+        check_success=True,
+        roof=True,
+        app_name="app_laplace_eqn",
+    )
+
+    # Analyze with FP32 data type
+    code_fp32 = binary_handler_analyze_rocprof_compute([
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Just an idea, you can provide multiple data types in single invocation and do one shot assertions for multiple data types

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 12]
```diff
@@ -3644,3 +3655,255 @@ def test_multi_rank_warning_pc_sampling(
...
+        "analyze",
+        "--path",
+        workload_dir,
+    ])
+    assert code in (0, 1)
+
+    test_utils.clean_output_dir(config["cleanup"], workload_dir)
+
+
+@pytest.mark.roofline_iteration_multiplexing
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Even though this is roofline specific, we can re-use `iteration_multiplexing_2` marker for this, no need to use a different marker for every test we add

File: projects/rocprofiler-compute/pyproject.toml [hunk 13]
```diff
@@ -111,6 +111,7 @@ markers = [
     "iteration_multiplexing_1",
     "iteration_multiplexing_2",
     "iteration_multiplexing_stochastic",
+    "roofline_iteration_multiplexing",
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/pyproject.toml]: Per my comments in test_profile_general.py we dont need this new marker

File: projects/rocprofiler-compute/CMakeLists.txt [hunk 14]
```diff
@@ -422,6 +422,14 @@ add_test(
         tests/test_profile_general.py ${WORKING_DIR_OPTION}
 )
 
+add_test(
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/CMakeLists.txt]: Per my comments in test_profile_general.py we dont need this new marker

**Review by vedithal-amd — CHANGES_REQUESTED** [review]
First review pass

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 2 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: Seconding this grouping- specifically having this in multiplexing section and not roofline functionality.

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 1 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: https://github.com/ROCm/rocm-systems/blob/932fb81cf9cc554f4d51210e2a205b29b2d7300e/projects/rocprofiler-compute/sample/laplace_eqn.hip#L186
agree, doesn't look like -o set to zero changes anything for kernel launch order. Does not affect our test case below using -o or not.

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 6 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: There is a comment above that asks to combine this check into the other iteration multiplexing tests- this could save us from having to profile multiple times, i.e. run roofline benchmarking again. We have issues with the pytest timeouts per bucket, so if this does not increase times combining the tests, great. Per this specific comment, if we are just reusing the same marker but adding more tests, we need to ensure that the timeouts are not being reached by adding more profiling runs to the bucket such that therock starts failing on us- same issue we have seen in the past requiring us to split up roofline markers.

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 9 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: Per the immediately above comments, this is not a roof-specific test. There is not a single kernel that has enough iterations, wouldn't the entire analysis fail? Or do we gracefully display zeroes/nans/something? More specifically, can we check for warning logs for insufficient passes from analysis_base.py::sanitize() if we gracefully continue with no complete data?

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 12 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: Same possible issue as above with the overloading the buckets to timeout, check if this is the case.

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 11 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: There may be an issue with CLI only being able to process one type at a time unlike GUI, but check if this works for block 4 except the plot (which will just take the first datatype arg and discard the rest)

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 5 above)
Comment (cfallows-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py:54]: Laplace is same kernel with different parameters. Using vcopy with multikernel option technically runs the same kernel an additional time but under a different name. At this time we don't have any iteration multiplexing test cases that have a workload that contains more than one kernel name- this would add that case to the mix.

**Suggestion: it would be helpful if we could add a control for indvidual iterations for vecCopy and vecCopy_2 into vcopy's args such that we could test for sufficient iterations for each as well as one/both of the kernels not having sufficient iterations.**

**Review by cfallows-amd — CHANGES_REQUESTED** [review]
A few comments but please see suggestion regarding control over the vcopy iterations change that we could implement, see if we can add a test case where there is a mix of sufficient and insufficient kernels.

Edit: I was having afterthoughts about testing using --iteration-multiplexing kernel and --iteration-multiplexing kernel_launch_params as well, but I think Abhinab might be covering this as well as dispatches vs bucket sizes (which I mentioned in my comment regarding vcopy multikernel) in another PR #3270

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 9 above)
Comment (vedithal-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: Per my knowledge, insufficient dispatches for iter. mplx. leads to graceful analysis using NaNs, we never crash abruptly.

Yes, capsys can be used to assert correct insufficient dispatches warning is shown in stdout

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 2 above)
Comment (jamessiddeley-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: Fair - will group the test cases together

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 5 above)
Comment (jamessiddeley-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py:54]: Yes I agree that insufficient dispatches test is a corner case that is relevant to the iter multiplexing and is covered by this unit test so we should keep it.

We can also modify the `vcopy` sample application source code to support per-kernel iteration control so we could test mixed sufficient/insufficient scenarios across different kernel names. Although, thats a source code change to a sample application not really in this scope as a test level change. I will follow up with on this to discuss adding this separately as it would be a positive addition.

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 12 above)
Comment (jamessiddeley-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: I've remove the pytest marker for `roofline_iteration_multiplexing` and now just re-using `iteration_multiplexing_2`

**Comment by jamessiddeley-amd** [comment]
Updated PR description to reflect changes addressing comments

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 15]
```diff
@@ -3165,6 +3218,96 @@ def test_iteration_multiplexing_all_counter_accuracy(
...
+        "kernels_with_missing_counters",
+        f"{workload_dir}/profiling_config.yaml",
+    )
+
+    code = binary_handler_analyze_rocprof_compute([
+        "analyze",
+        "--path",
+        workload_dir,
+    ])
+    assert code == 0
```
Comment (abchoudh-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py:3252]: Add an assertion that the warning for insufficient kernels is printed.

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 16]
```diff
@@ -3165,6 +3218,96 @@ def test_iteration_multiplexing_all_counter_accuracy(
...
+    )
+
+    code_fp32 = binary_handler_analyze_rocprof_compute([
+        "analyze",
+        "--path",
+        workload_dir,
+        "--roofline-data-type",
+        "FP32",
+    ])
+    assert code_fp32 == 0
```
Comment (abchoudh-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py:3299]: Add assertion for the data existence.

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 17]
```diff
@@ -3165,6 +3218,96 @@ def test_iteration_multiplexing_all_counter_accuracy(
...
+
+    code_fp16 = binary_handler_analyze_rocprof_compute([
+        "analyze",
+        "--path",
+        workload_dir,
+        "--roofline-data-type",
+        "FP16",
+    ])
+    assert code_fp16 == 0
```
Comment (abchoudh-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py:3308]: Sane here.

**Review by abchoudh-amd — CHANGES_REQUESTED** [review]

**Review by abchoudh-amd — APPROVED** [review]

**Review by xuchen-amd — APPROVED** [review]

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 18]
```diff
@@ -3021,39 +3035,56 @@ def test_iteration_multiplexing_deterministic_counter_accuracy(
...
+        assert os.path.exists(f"{workload_dir_klp}/roofline.csv")
+        roofline_df = pd.read_csv(f"{workload_dir_klp}/roofline.csv")
+        assert len(roofline_df) >= num_devices
+
+        code = binary_handler_analyze_rocprof_compute([
+            "analyze",
+            "--path",
+            workload_dir_klp,
+        ])
+        assert code == 0
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Let's move it outside if block and test again.
Even without roofline.csv - analysis should still gracefully handle it and exit with code 0

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 19]
```diff
@@ -3084,29 +3122,44 @@ def test_iteration_multiplexing_stochastic_counter_accuracy(
...
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: Same as above, move out of if block

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 20]
```diff
@@ -3163,6 +3216,104 @@ def test_iteration_multiplexing_all_counter_accuracy(
...
+    binary_handler_profile_rocprof_compute,
+    binary_handler_analyze_rocprof_compute,
+    capsys,
+):
+    """Verify graceful degradation when dispatches are too few for full
+    counter coverage under iteration multiplexing.
+    """
+    options = [
+        "--block",
+        "4",
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: We have already tested roofline with iter. mplx. above.

This test is iter. mplx. specific insufficient dispatches handling test.

Do we really need roofline counters here?
If you agree, please remove block=4 and roof=True
This will also help minimize testing time which is already a lot.

File: projects/rocprofiler-compute/tests/test_profile_general.py [hunk 21]
```diff
@@ -3163,6 +3216,104 @@ def test_iteration_multiplexing_all_counter_accuracy(
...
+    on iteration-multiplexed profiling data.
+    """
+    if soc in ("MI100"):
+        pytest.skip("Roofline not supported on MI100")
+        return
+
+    options = [
+        "--block",
+        "4",
+        "6.1.5",
```
Comment (vedithal-amd) [inline projects/rocprofiler-compute/tests/test_profile_general.py]: 6.1.5 block can be removed

**Review by vedithal-amd — APPROVED** [review]
Minor comments, looks good

File: projects/rocprofiler-compute/tests/test_profile_general.py (hunk 20 above)
Comment (jamessiddeley-amd) [reply projects/rocprofiler-compute/tests/test_profile_general.py]: I agree thats a good point. Roofline coverage is already handled by the deterministic and stochastic tests and the test case is more of an iteration multiplexing concern not a roofline concern

**Review by cfallows-amd — APPROVED** [review]
